### ISSUE TITLE
Support for `Optional`

### ISSUE THREAD

**Issue by prdoyle** [google/gson#1102]
Usually, when I use Gson, I need to add special logic so that a field of `Optional<X>` gets serialized as `X` if it's present, and gets omitted if it's empty; and conversely on deserialization.  This seems like the only reasonable way to handle `Optional`.  Would you consider offering this feature (or accepting a PR that implements it)?

**Comment by JakeWharton** [comment]
Gson can't depend on Java 8 types as it supports down to Java 6.

**Comment by sspies8684** [comment]
I would really like to see this.

**Comment by kaqqao** [comment]
@JakeWharton So? It could always be in a separate module.

**Comment by stevenschlansker** [comment]
It looks like such a module exists: https://github.com/hsiafan/gson-java8-datatype
The errors emitted if you forget to include it are very confusing and do not obviously lead you to determining that you must install this module, but once you do, `Optional` types work fine.

**Comment by marcblanchet** [comment]
really would like to see this. Coming from Swift, the use of Optional with JSON is very useful.

**Comment by Marcono1234** [comment]
> Gson can't depend on Java 8 types as it supports down to Java 6.

With [Multi-Release JARs](https://openjdk.java.net/jeps/238) this would now be possible, at least for JRE >= 9.

**Comment by mleonhard** [comment]
I need this.

**Comment by chakpongchung** [comment]
I need this

**Comment by kaqqao** [comment]
@mleonhard @chakpongchung Did you guys read the thread? What you want is already there, use it: https://github.com/hsiafan/gson-java8-datatype

**Comment by piegamesde** [comment]
@kaqqao gson-java8-datatype worked great, but sadly it broke with Java 16 and isn't maintained anymore (the linked repository is archived).

> Gson can't depend on Java 8 types as it supports down to Java 6.

Java 6 is now end of life for 8(!) years, and Java 8 (and thus, Optional) exist for 7 years now. I think it's about time to move on.

**Comment by alx741** [comment]
I agree that keeping that kind of retro-compatibility over including new useful features for this long might be unreasonable. Java 8 has been around for a long time now

**Comment by prdoyle** [comment]
What is the value of supporting Java versions past their end of life?

**Comment by eamonnmcmanus** [comment]
I think we could support `Optional` in a similar way to the support for `java.time` added in #2948. That is, we add code that explicitly serializes and deserializes into the format that previously would have been obtained through reflection on the internal fields of `java.util.Optional`. This is probably the best we can do while still preserving compatibility. Like the `java.time` code, Gson would access this new code through reflection so it can still remain compatible with Java 7.

Concretely, the serial forms would look like this:

Java | Json
------|-----
`Optional.of("foo")` | `{"value":"foo"}`
`Optional.of(23)` | `{"value":23}`
`Optional.empty()` | `{}`
`OptionalInt.of(23)`  | `{"isPresent":true,"value":23}`
`OptionalInt.empty()` | `{"isPresent":false,"value":0}`

and similarly for `OptionalDouble` and `OptionalLong`. The encoding of `Optional` is not too terrible, though the others kind of are.

Like `java.time`, this is not the encoding we would have chosen for these types in the absence of compatibility concerns. It might be worthwhile to supply some adapters that are not installed in `Gson` instances by default, but that can be added manually. For `java.time`, that would typically use the ISO standard encodings of dates and times, and for `Optional*` it would probably just encode `Optional.of("foo")` as `"foo"` and `Optional.empty()` as `null`.

**Comment by Marcono1234** [comment]
> Like the `java.time` code, Gson would access this new code through reflection so it can still remain compatible with Java 7.

With the current Gson code on `main` requiring >= Java 8 and >= Android API level 24, it can probably be implemented without any need for reflection because [`java.util.Optional` is available since API level 24](https://developer.android.com/reference/java/util/Optional).
